kong/plugins/moesif/moesif_gov.lua

```diff
@@ -0,0 +1,310 @@
+
+                local updated_rule_values = {}
+                if rule_values ~= nil then 
+                    for k,v in pairs(rule_values) do updated_rule_values["{{"..k.."}}"]=v end
+                end
+
+                -- Check if rule_values is table and not empty
+                if rule_values ~= nil and type(updated_gr_headers) == "table" and next(updated_gr_headers) ~= nil then
+                    -- Replace headers
+                    for headerName, headerValue in pairs(updated_gr_headers) do
+                        updated_gr_headers[headerName] = headerValue:gsub("{{%d+}}", updated_rule_values)
+                    end
+                    -- Replace body
+                    updated_gr_body = transformValues(updated_gr_body, updated_rule_values)
+                end 
+                -- Add blocked_by field to the event to determine the rule by which the event was blocked
+                conf["blocked_by"] = rule_id
+
+                local end_access_phase_time = socket.gettime()*1000
+                ngx.log(ngx.DEBUG, "[moesif] access phase took time for blocking request - ".. tostring(end_access_phase_time - start_access_phase_time))
+
+                return kong.response.exit(gr_status, updated_gr_body, updated_gr_headers)
+            else 
+                if conf.debug then
+                    ngx_log(ngx.DEBUG, "[moesif] Skipped blocking request as governance rule response is not set for the entity Id -  "..entity_id)
+                end
+                return nil
+            end
+        else 
+            if conf.debug then
+                ngx_log(ngx.DEBUG, "[moesif] Skipped blocking request as no governance rule found or for none of the governance rule block is set to true for the entity Id - "..entity_id)
+            end
+            return nil
+        end
+    else
+        if conf.debug then 
+            ngx_log(ngx.DEBUG, "[moesif] Skipped blocking request as Entity rules are empty or no governance rules defined for entity Id - "..entity_id)
+        end
+        return nil
+    end
+end
+
+function _M.govern_request(ngx, conf, start_access_phase_time)
+
+    -- Hash key of the config application Id
+    local hash_key = ngx_md5(conf.application_id)
+    local user_id_entity
+    local company_id_entity
+    local request_headers = req_get_headers()
+
+    -- Set fetch_governance_rule to true (default), to be able to fetch governance rule
+    if fetch_governance_rules[hash_key] == nil then
+        fetch_governance_rules[hash_key] = true
+    end
+
+    -- Fetch the governance rules
+    if (governance_rules_etags[hash_key] == nil or (conf["rulesETag"] ~= governance_rules_etags[hash_key])) and fetch_governance_rules[hash_key] then
+        -- Set fetch_governance_rule to false, to avoid fetching governance rule
+        fetch_governance_rules[hash_key] = false
+
+        local gr_ok, gr_err = ngx_timer_at(0, get_governance_rules, hash_key, conf)
```

Comment:
why are we fetching governance rules here? can we not do in init, i think its worker process start something.. its not correct to start timer here i think

Reply:
govern_request this looks to be fcn that gets called on every request? correct me if i m wrong

Reply:
I dont like us getting governance rules in sync to kong request, even if governance rule_etags is empty. Simply return from here if config is empty.